src/main/java/com/codeborne/selenide/impl/DownloadFileToFolderCdp.java

```diff
@@ -0,0 +1,154 @@
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.codeborne.selenide.impl.FileHelper.moveFile;
+import static java.lang.System.currentTimeMillis;
+import static java.lang.Thread.sleep;
+import static org.openqa.selenium.devtools.v119.browser.model.DownloadProgress.State.COMPLETED;
+
+@ParametersAreNonnullByDefault
+public class DownloadFileToFolderCdp {
+  private static final Logger log = LoggerFactory.getLogger(DownloadFileToFolderCdp.class);
+  private final Downloader downloader;
+  private AtomicBoolean downloadComplete = new AtomicBoolean(false);
+  private AtomicReference<String> fileName = new AtomicReference<>();
+
+  private DevTools devTools;
+
+  DownloadFileToFolderCdp(Downloader downloader) {
+    this.downloader = downloader;
+  }
+
+  public DownloadFileToFolderCdp() {
+    this(new Downloader());
+  }
+
+  @CheckReturnValue
+  @Nonnull
+  public File download(WebElementSource anyClickableElement,
+                       WebElement clickable, long timeout,
+                       FileFilter fileFilter,
+                       DownloadAction action) {
+
+    return clickAndWaitForNewFilesInDownloadsFolder(anyClickableElement, clickable, timeout, fileFilter, action);
+  }
+
+  @CheckReturnValue
+  @Nonnull
+  private File clickAndWaitForNewFilesInDownloadsFolder(WebElementSource anyClickableElement, WebElement clickable,
+                                                        long timeout,
+                                                        FileFilter fileFilter,
+                                                        DownloadAction action) {
+    Driver driver = anyClickableElement.driver();
+    Config config = driver.config();
+    long pollingInterval = Math.max(config.pollingInterval(), 50);
+    DownloadsFolder folder = getDownloadsFolder(driver);
+
+    if (folder == null) {
+      throw new IllegalStateException("Downloads folder is not configured");
+    }
+
+    folder.cleanupBeforeDownload();
+    prepareDownloadWithCdp(driver);
+
+    long downloadStartedAt = currentTimeMillis();
+
+    action.perform(driver, clickable);
+
+    waitUntilDownloadsCompleted(timeout, pollingInterval);
+
+    Downloads newDownloads = new Downloads(folder.filesNewerThan(downloadStartedAt));
+    if (log.isInfoEnabled()) {
+      log.info("Downloaded files in {}: {}", folder, newDownloads.filesAsString());
+    }
+    if (log.isDebugEnabled()) {
+      log.debug("All downloaded files: {}", folder.filesAsString());
+    }
+
+    File downloadedFile = newDownloads.firstDownloadedFile(driver, timeout, fileFilter);
+    return archiveFile(driver, downloadedFile);
+  }
+
+  @Nullable
+  protected DownloadsFolder getDownloadsFolder(Driver driver) {
+    return driver.browserDownloadsFolder();
+  }
+
+  private void waitUntilDownloadsCompleted(long timeout, long pollingInterval) {
+    for (long start = currentTimeMillis(); currentTimeMillis() - start < timeout; pause(pollingInterval)) {
```

Comment:
Please use `com.codeborne.selenide.Stopwatch` for polling wait needs. See example usage in `com.codeborne.selenide.impl.SelenideElementProxy#dispatchAndRetry`

Reply:
Fixed